### PR TITLE
Deprecate position argument to Axes.apply_aspect()

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

In itself `apply_aspect()` feels too low level to be needed by an end-user. I'm not sure if it reasonably works at all on its own. It might be that the result is influenced by other artists and if something is left stale, then the application result is still not final. In that case, one would need a full draw (without rendering) anyway. OTOH if `apply_aspect()` was really standalone, I don't know why we defer the application and don't apply it immediately when setting the aspect.

Also, the *position* parameter of `apply_aspect` does not seem a user-facing API. It is used internally by layout mechanisms. But shifting the position explicitly as part of aspect application seems overreach of the scope of the function and at least not something a user should do. 

I'm 99% percent sure `apply_aspect()` does not need to be public API, but please speak up if you think otherwise. And even if we miss a justified use-case for a user. They should be able to give feedback during the deprecation period and we would reconsider then.

### CODE REVIEW

**Review by tacaswell — APPROVED** [review]

File: lib/matplotlib/axes/_base.py [hunk 1]
```diff
@@ -1855,7 +1855,11 @@ def get_data_ratio(self):
         ysize = max(abs(tymax - tymin), 1e-30)
         return ysize / xsize
 
+    @_api.deprecated("3.6", alternative="Figure.draw_without_rendering")
     def apply_aspect(self, position=None):
+        self._apply_aspect(position)
```
Comment (QuLogic) [inline lib/matplotlib/axes/_base.py:1860]: Could do with a test.

File: lib/matplotlib/axes/_base.py [hunk 2]
```diff
@@ -1855,7 +1855,11 @@ def get_data_ratio(self):
...
+
+    def _apply_aspect(self, position=None):
         """
         Adjust the Axes for a specified data aspect ratio.
```
Comment (QuLogic) [inline lib/matplotlib/axes/_base.py:1864]: Shouldn't the docstring stay on the public method until it's gone?

File: lib/mpl_toolkits/axes_grid1/parasite_axes.py [hunk 3]
```diff
@@ -82,9 +82,13 @@ def _update_viewlim(self):  # Inline after deprecation elapses.
         else:
             _api.check_in_list([None, "equal", "transform"], mode=mode)
 
+    @_api.deprecated("3.6", alternative="Figure.draw_if_interactive")
     def apply_aspect(self, position=None):
+        self._apply_aspect(position)
```
Comment (QuLogic) [inline lib/mpl_toolkits/axes_grid1/parasite_axes.py:87]: Also could do with a test.

**Comment by jklymak** [comment]
I think if you search on stackoverflow there are quite a few answers that suggest calling apply_aspect to get the position of things after the proper aspect ratio is applied without forcing a full draw.

**Comment by jklymak** [comment]
Well, maybe only a couple:  https://stackoverflow.com/search?q=%5Bmatplotlib%5D+apply_aspect

**Comment by timhoffm** [comment]
> Well, maybe only a couple: https://stackoverflow.com/search?q=%5Bmatplotlib%5D+apply_aspect

From these:
- 3) and 5) only show `apply_aspect` in unrelated parts of a traceback.
- 1), 2), 4) use it for a layout update because they want to query Axes sizes. - Note also that in 1  @tacaswell suggests "I strongly suspect that you just need to call draw after calling set_aspect."

Anyway, let's be super-safe and only deprecate the public *position* parameter for now. - No end user should manipulate the position through `apply_aspect()`. And I haven't seen anything in the wild.

**Comment by QuLogic** [comment]
Please update the commit message as well for that.

File: doc/api/next_api_changes/deprecations/23629-TH.rst [hunk 4]
```diff
@@ -0,0 +1,3 @@
+The parameter *position* in ``Axes.apply_aspect()``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+... is deprecated. It is considered internal API and now public use case is known.
```
Comment (QuLogic) [inline doc/api/next_api_changes/deprecations/23629-TH.rst:3]:
```suggestion
... is deprecated. It is considered internal API and no public use case is known.
```

File: lib/matplotlib/axes/_base.py [hunk 5]
```diff
@@ -1866,10 +1867,17 @@ def apply_aspect(self, position=None):
         Parameters
         ----------
         position : None or .Bbox
+
             If not ``None``, this defines the position of the
```
Comment (QuLogic) [inline lib/matplotlib/axes/_base.py:1871]:
```suggestion
        position : None or .Bbox
            If not ``None``, this defines the position of the
```

**Review by QuLogic — COMMENTED** [review]
I've found a use of `apply_aspect` in the wild that this rename would likely break without the deprecation triggering
https://github.com/AudioSceneDescriptionFormat/splines/blob/master/doc/rotation/helper.py#L275

I'm not entirely sure what they are attempting to do there.

**Comment by QuLogic** [comment]
One use of `apply_aspect` that maybe could be a `draw`: https://github.com/hannorein/rebound/blob/main/rebound/plotting.py#L103-L107

And one that passes a position: https://github.com/astropy/astropy/blob/main/astropy/visualization/wcsaxes/core.py#L444-L452

**Comment by timhoffm** [comment]
Deferred to 3.7. Let’s not rush this into 3.6.

**Comment by timhoffm** [comment]
Digging through the history my understanding of the reason for the position parameter is the following:

It is somewhat surprising that "_apply_aspect" takes an optional position as input, which seems more functionality than what the name suggests.

The parameter was introduced in https://github.com/matplotlib/matplotlib/commit/66290aae. Generally, applying an aspect will modify the size and position of an Axes. The fact that position is updated anyway was used to funnel additional position information from a layoutinto the already existing positioning code.

IMHO:
- The *position* parameter should never be used by an end-user.
- The function `_apply_aspect` should rather be called `update_geometry` or similar.

I see two ways forward:
1. Simply document that *position* is not meant to be user-facing. And don't bother further with the topic.
2. Go further with a formal deprecation, which means we need a public `apply_aspect()` that will become parameter-less. And we need a new private API to carry the position. This has compatibility implications for 3rd party Axes subclasses (see @QuLogic's links above).

I'm inclined to go with 1.

---

> One use of apply_aspect that maybe could be a draw: https://github.com/hannorein/rebound/blob/main/rebound/plotting.py#L103-L107

has meanwhile been changed to a draw. 👍

**Pull request closed without merging**